File: Java/49_Group_Anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> sol = new ArrayList<>();
        HashMap<String, List<String>> map = new HashMap<>();

        for(String s : strs) {

            int [] curr = getArray(s);
            String convertedArray = arrayToString(curr);
            if(map.containsKey(convertedArray)) {
                map.get(convertedArray).add(s);
            } else {
                List<String> newList = new ArrayList<>();
                newList.add(s);
                map.put(convertedArray, newList);
            }

        }
        for(List<String> currList : map.values()) {
            sol.add(currList);
        }
        return sol;
    }

    private int [] getArray(String str) {
        int [] sol = new int [26];
        for(char c : str.toCharArray()) {
            sol[c - 'a']++;
        }
        return sol;
    }

    private String arrayToString(int [] arr) {
        StringBuilder sb = new StringBuilder();
        for(int i = 0; i < arr.length;i++) {
            sb.append(arr[i]).append(',');
        }
        return sb.toString();
    }



    
}

File: Python/49_Group_Anagrams.py
class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        def getAnagram(str):
            arr = [0] * 26
            for i in range(len(str)):
                arr[ord(str[i]) - ord('a')] += 1
            return tuple(arr)

        anagrams = {}

        for i in range(len(strs)):
            key = getAnagram(strs[i])

            if key not in anagrams:
                anagrams[key] = []
            anagrams[key].append(strs[i])
        
        return list(anagrams.values())

        
